demo/midgard-node/src/transactions/utils.ts

```diff
@@ -50,19 +76,50 @@ export const handleSignSubmit = (
  */
 export const handleSignSubmitWithoutConfirmation = (
   signBuilder: TxSignBuilder,
-): Effect.Effect<string, Error> =>
+  onSubmitFailure: (error: SubmitError) => Effect.Effect<void>,
+): Effect.Effect<string | void, Error> =>
   Effect.gen(function* () {
-    const signed = yield* signBuilder.sign.withWallet().completeProgram();
-    const txHash = yield* signed
-      .submitProgram()
-      .pipe(
-        Effect.retry(
-          Schedule.compose(Schedule.exponential(5_000), Schedule.recurs(5)),
-        ),
-      );
+    const signed = yield* signBuilder.sign
+      .withWallet()
+      .completeProgram()
+      .pipe(Effect.mapError((err) => new SignError({ err })));
+    const txHash = yield* signed.submitProgram().pipe(
+      Effect.retry(
+        Schedule.compose(Schedule.exponential(5_000), Schedule.recurs(5)),
+      ),
+      Effect.mapError((err) => new SubmitError({ err })),
+    );
     yield* Effect.logDebug(`🚀 Transaction submitted: ${txHash}`);
     return txHash;
-  });
+  }).pipe(
+    Effect.catchAll((err: SignError | SubmitError) =>
+      Effect.gen(function* () {
+        switch (err._tag) {
+          case "SubmitError":
+            onSubmitFailure(err);
+            break;
+          case "SignError":
+            yield* Effect.logError(`Signing tx error: ${err.err}`);
+            return yield* Effect.fail(err);
+        }
+      }),
```

Comment:
Same as the other comment. You can also avoid `Effect.gen` by using `pipe` on `"SignError"`.